### PR TITLE
NNVP-74,NNVP-75,NNVP-76

### PR DESCRIPTION
**Pull request by mateusztheglue**

### CODE REVIEW

**Comment by laurent-theglue** [comment]
:star: PR built and available in a preview environment **the-glue-nnvp-backend-pr-37** [here](http://nnvp-backend.jx-the-glue-nnvp-backend-pr-37.cluster.development.nnvp.thegluesandbox.com)

File: src/main/java/com/theglue/common/repository/search/SearchHelper.java [hunk 1]
```diff
@@ -87,16 +87,16 @@ public SearchHelper(
...
     Root<T> criteriaQueryRoot = criteriaQuery.from(rootClass);
     criteriaQuery
-        .select(criteriaQueryRoot.get(projectionField))
+        .multiselect(projectionFields.stream().map(criteriaQueryRoot::get).collect(Collectors.toList()))
         .where(getPredicates(parameters, criteriaBuilder, criteriaQuery, criteriaQueryRoot))
-        .orderBy(criteriaBuilder.asc(criteriaQueryRoot.get(projectionField)))
+        .orderBy(projectionFields.stream().map(criteriaQueryRoot::get).map(criteriaBuilder::asc).collect(Collectors.toList()))
         .distinct(true);
     return em.createQuery(criteriaQuery).getResultList();
   }
```
Comment (eryk7b) [inline src/main/java/com/theglue/common/repository/search/SearchHelper.java]: Just a not important idea for the formatting. Streams in seperate lines to have IDEA both higlight on what we're working on and to know which line throws if there's any Exception
```
  public <P> List<P> searchWithProjection(List<SearchParameter> parameters, List<String> projectionFields, Class<P> projectionClass) {
    CriteriaBuilder criteriaBuilder = em.getCriteriaBuilder();

    CriteriaQuery<P> criteriaQuery = criteriaBuilder.createQuery(projectionClass);
    Root<T> criteriaQueryRoot = criteriaQuery.from(rootClass);
    criteriaQuery
        .multiselect(projectionFields.stream()
            .map(criteriaQueryRoot::get)
            .collect(Collectors.toList())
        )
        .where(getPredicates(parameters, criteriaBuilder, criteriaQuery, criteriaQueryRoot))
        .orderBy(projectionFields.stream()
            .map(criteriaQueryRoot::get)
            .map(criteriaBuilder::asc)
            .collect(Collectors.toList())
        )
        .distinct(true);
    return em.createQuery(criteriaQuery).getResultList();
```

File: src/main/java/com/theglue/nnvp/config/JacksonConfig.java [hunk 2]
```diff
@@ -11,10 +12,12 @@
 
   @Bean
   public ObjectMapper objectMapper() {
-    return new ObjectMapper()
+    ObjectMapper mapper = new ObjectMapper()
         .findAndRegisterModules()
         .setSerializationInclusion(JsonInclude.Include.NON_NULL)
         .setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE);
+    mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, true);
+    return mapper;
```
Comment (eryk7b) [inline src/main/java/com/theglue/nnvp/config/JacksonConfig.java]: didn't just .configure work?
```
    return new ObjectMapper()
        .findAndRegisterModules()
        .setSerializationInclusion(JsonInclude.Include.NON_NULL)
        .setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE)
        .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, true);
```
but again it doesn't matter that much

File: src/main/java/com/theglue/nnvp/controller/VehicleProjectionSearchController.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+  @GetMapping(path = "/vehicle/makes")
+  @ResponseBody
+  @ApiOperation(value = "Get vehicle makes")
+  public ResponseEntity<SuccessResponseDTO<GetVehicleMakesResponseDTO>> getVehicleMakes(@Valid GetVehicleMakesRequestDTO request) {
+
+    List<SearchParameter> searchParameters = SearchRequestUtils.toSearchParameters(request,
+        "vehicleCategoryList", "language", "year", "uniqueIdentifier");
+    enrichWithVehicleCategoryParameter(searchParameters, request.getVehicleCategoryList());
+    enrichWithConstructionYearParameter(searchParameters, request.getYear());
```
Comment (eryk7b) [inline src/main/java/com/theglue/nnvp/controller/VehicleProjectionSearchController.java]: probably we need a change from construction year naming into some manufacturing name (here and in method parameter definition etc)

File: src/main/java/com/theglue/nnvp/controller/VehicleProjectionSearchController.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    return new ResponseEntity<>(
+        SuccessResponseDTO.successResponse(response),
+        HttpStatus.OK
+    );
+  }
+
+  @GetMapping(path = "/vehicle/types")
+  @ResponseBody
+  @ApiOperation(value = "Get vehicle types")
+  public ResponseEntity<SuccessResponseDTO<GetVehicleTypesResponseDTO>> getVehicleModels(@Valid GetVehicleTypesRequestDTO request) {
```
Comment (eryk7b) [inline src/main/java/com/theglue/nnvp/controller/VehicleProjectionSearchController.java]: getVehicleTypes probably

**Comment by laurent-theglue** [comment]
:star: PR built and available in a preview environment **the-glue-nnvp-backend-pr-37** [here](http://nnvp-backend.jx-the-glue-nnvp-backend-pr-37.cluster.development.nnvp.thegluesandbox.com)

**Comment by laurent-theglue** [comment]
:star: PR built and available in a preview environment **the-glue-nnvp-backend-pr-37** [here](http://nnvp-backend.jx-the-glue-nnvp-backend-pr-37.cluster.development.nnvp.thegluesandbox.com)

**Pull request merged**